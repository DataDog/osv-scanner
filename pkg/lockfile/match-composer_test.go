package lockfile_test

import (
	"io/fs"
	"os"
	"path/filepath"
	"testing"

	"github.com/google/osv-scanner/pkg/models"

	"github.com/google/osv-scanner/internal/testutility"

	"github.com/google/osv-scanner/pkg/lockfile"
	"github.com/stretchr/testify/assert"
)

var composerMatcher = lockfile.ComposerMatcher{}

func TestComposerMatcher_GetSourceFile_FileDoesNotExist(t *testing.T) {
	t.Parallel()

	lockFile, err := lockfile.OpenLocalDepFile("fixtures/composer/no-json/composer.lock")
	if err != nil {
		t.Errorf("Got unexpected error: %v", err)
	}

	sourceFile, err := composerMatcher.GetSourceFile(lockFile)
	expectErrIs(t, err, fs.ErrNotExist)
	assert.Equal(t, "", sourceFile.Path())
}

func TestComposerMatcher_GetSourceFile(t *testing.T) {
	t.Parallel()
	dir, err := os.Getwd()
	if err != nil {
		t.Errorf("Got unexpected error: %v", err)
	}

	basePath := "fixtures/composer/one-package/"
	sourcefilePath := filepath.FromSlash(filepath.Join(dir, basePath+"composer.json"))

	lockFile, err := lockfile.OpenLocalDepFile(basePath + "composer.lock")
	if err != nil {
		t.Errorf("Got unexpected error: %v", err)
	}

	sourceFile, err := composerMatcher.GetSourceFile(lockFile)
	if err != nil {
		t.Errorf("Got unexpected error: %v", err)
	}

	assert.Equal(t, sourcefilePath, sourceFile.Path())
}

func TestComposerMatcher_Match_OnePackage(t *testing.T) {
	t.Parallel()

	sourceFile, err := lockfile.OpenLocalDepFile("fixtures/composer/one-package/composer.json")
	if err != nil {
		t.Errorf("Got unexpected error: %v", err)
	}

	packages := []lockfile.PackageDetails{
		{
			Name:           "brick/math",
			Version:        "0.12.9",
			PackageManager: models.Composer,
			IsDirect:       true,
		},
	}
	err = composerMatcher.Match(sourceFile, packages)
	if err != nil {
		t.Errorf("Got unexpected error: %v", err)
	}

	testutility.NewSnapshot().MatchText(t, testutility.NormalizeJSON(t, packages))
}
