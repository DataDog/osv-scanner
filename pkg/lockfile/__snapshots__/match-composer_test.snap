
[TestComposerMatcher_Match_OnePackage - 1]
[
  {
    "name": "brick/math",
    "version": "0.12.9",
    "blockLocation": {
      "line": {
        "start": 18,
        "end": 19
      },
      "column": {
        "start": 1,
        "end": 35
      },
      "file_name": "<rootdir>/fixtures/composer/one-package/composer.json"
    },
    "versionLocation": {
      "line": {
        "start": 19,
        "end": 19
      },
      "column": {
        "start": 8,
        "end": 34
      },
      "file_name": "<rootdir>/fixtures/composer/one-package/composer.json"
    },
    "nameLocation": {
      "line": {
        "start": 18,
        "end": 18
      },
      "column": {
        "start": 6,
        "end": 16
      },
      "file_name": "<rootdir>/fixtures/composer/one-package/composer.json"
    }
  }
]
---

[TestComposerMatcher_OnePackage_MatcherFailed - 1]
[
  {
    "name": "brick/math",
    "version": "0.12.1",
    "commit": "f510c0a40911935b77b86859eb5223d58d660df1",
    "ecosystem": "Packagist",
    "compareAs": "Packagist",
    "blockLocation": {
      "line": {
        "start": 0,
        "end": 0
      },
      "column": {
        "start": 0,
        "end": 0
      },
      "file_name": ""
    }
  }
]
---
